### PR TITLE
Add shallow api

### PR DESCRIPTION
**Pull request by mzgoddard**
Following the direction in #9, add a `shallow(vdom)` export that returns a shallow rendered dom instead of the deeper dom returned by `render(vdom)`.

- Implement `attrs()`

### CODE REVIEW

File: src/preact-render-spy.js [hunk 1]
```diff
@@ -5,70 +5,105 @@ const {isWhere} = require('./is-where');
...
   if (!vdom) {
     return vdom;
   }
   if (typeof vdom.nodeName === 'function' && !vdom.nodeName.isSpy) {
     vdom = Object.assign({}, vdom, {
-      nodeName: createSpy(spy, vdom.nodeName),
+      nodeName: (spy._createSpy && spy._createSpy(vdom, depth) || createSpy)(spy, vdom.nodeName),
       attributes: Object.assign({}, vdom.attributes, {
         [SPY_PRIVATE_KEY]: vdom,
+        [SPY_PRIVATE_DEPTH]: depth,
```
Comment (gnarf) [inline src/preact-render-spy.js]: Why not store `[SPY_PRIVATE_KEY]: {vdom, depth}` ? I feel like we should limit the number of keys we use on this spy vdom

File: src/preact-render-spy.js [hunk 2]
```diff
@@ -5,70 +5,105 @@ const {isWhere} = require('./is-where');
 const {selToWhere} = require('./sel-to-where');
 
 const SPY_PRIVATE_KEY = 'SPY_PRIVATE_KEY';
```
Comment (gnarf) [inline src/preact-render-spy.js]: I feel like we should:

```js
export const config = {
  SPY_PRIVATE_KEY: 'SPY_PRIVATE_KEY',
};
```

just in case someone wants to:

```js
import {config} from 'preact-render-spy';
config.SPY_PRIVATE_KEY='@@somethingelse';

File: src/preact-render-spy.js [hunk 3]
```diff
@@ -176,14 +238,25 @@ class FindWrapper {
...
+  /**
    * Return the text of all nested children concatenated together.
    */
   text() {
     return Array.from(vdomWalk(this.spy.vdomMap, Array.from(this)))
     // Filter for strings (text nodes)
     .filter(value => typeof value === 'string')
     // Concatenate all strings together
-    .reduce((carry, value) => carry + value);
+    .reduce((carry, value) => carry + value, '');
```
Comment (gnarf) [inline src/preact-render-spy.js]: Why not `.join('')` ?

File: src/preact-render-spy.js [hunk 4]
```diff
@@ -209,8 +282,11 @@ class FindWrapper {
...
-const renderSpy = vdom => {
-  return new SpyWrapper().render(vdom);
-};
+const renderDeep = vdom => new DeepWrapper().render(vdom);
 
-module.exports = renderSpy;
+const shallow = vdom => new ShallowWrapper().shallow(vdom);
+
+module.exports = {
+  render: renderDeep,
```
Comment (gnarf) [inline src/preact-render-spy.js]: maybe we should just call this `deep`? shrug

File: src/preact-render-spy.js [hunk 5]
```diff
@@ -99,15 +136,40 @@ class SpyWrapper {
     return new FindWrapper(this, [this.vdomMap.get('root')], selector);
   }
 
-  render(vdom) {
+  _createSpy(vdom, depth) {
+    return createSpy;
+  }
+
+  __render(vdom) {
```
Comment (gnarf) [inline src/preact-render-spy.js]: since we don't really change anything about this method between the two wrappers, perhaps we should just call it `render` here and on the shallow wrapper?

`new DeepWrapper().render(jsx)` vs `new ShallowWrapper.render(jsx)`

File: src/shared-render.test.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
+const {h, Component} = require('preact');
+
+const {render, shallow} = require('./preact-render-spy');
+
+const shape = (func, funcMember) => {
```
Comment (gnarf) [inline src/shared-render.test.js]: `sharedTests` maybe?  This doesn't feel like "shape"

Comment (gnarf) [inline src/shared-render.test.js]: also, you could probably just pass `funcMember` and destructure `{[funcMember]: func}` from the require:

```js

const renderSpy = require('./preact-render-spy');
const sharedTests = method => {
  const { [method]: func } = renderSpy;
};

File: src/shared-render.test.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  });
+
+  it(`${funcMember}: renders stateless components`, () => {
+    const Node = ({count}) => <div>{count}</div>;
+    const context = func(<Node count="1" />);
+    expect(context.find('div').length).toBe(1);
+    expect(context.find('div').contains('1')).toBeTruthy();
+  });
+
+  it(`${funcMember}: renders components with null children`, () => {
```
Comment (gnarf) [inline src/shared-render.test.js]: not that it's a big deal, but since `shallow` wouldn't ever even render `Empty` I think we should potentially add a test where just `Node = () => null;` and `context = func(<div><Node />text</div>)`

File: src/shallow-render.test.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+const {h, Component} = require('preact');
+
+const {shallow} = require('./preact-render-spy');
+
+it('shallow renders the first level of Components', () => {
+  const Node = ({children}) => <div>{children}</div>;
+  const context = shallow(<Node><Node /></Node>);
+  expect(context.find('div').length).toBe(1);
```
Comment (gnarf) [inline src/shallow-render.test.js]: this expect is a bit confusing, but I get it...  it brings up some interesting alternate cases though.


* `shallow(<span><Node /></span>)` should still render the `Node` yeah?
* Should we add a test for `const ErrorIfRendered = () => throw new Error('should not ever render'), Node = () => <ErrorIfRendered />, context = shallow(<Node />)` to really ensure that the child component is never instantiated?

File: src/preact-render-spy.js (hunk 3 above)
Comment (mzgoddard) [inline src/preact-render-spy.js]: This fixes a bug. `[].reduce((c, v) => c + v)` produces an error. If you reduce an empty array reduce doesn't have a deterministic value to return. For possibly empty arrays you must start with a value.

File: src/shared-render.test.js (hunk 6 above)
Comment (mzgoddard) [inline src/shared-render.test.js]: I renamed with `sharedTests`. Changed the args to `name, func`. `name` to add something to the test name for debugging failures and `func` to pass the func to use in.

File: src/preact-render-spy.js (hunk 3 above)
Comment (gnarf) [inline src/preact-render-spy.js]: Right, but isn't this just join('') ?

Comment (mzgoddard) [inline src/preact-render-spy.js]: Lol. Yes, you're totally correct.
